contracts/extensions/ExtensionManager.sol

```diff
@@ -0,0 +1,167 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with The Colony Network. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+pragma solidity 0.5.8;
+pragma experimental ABIEncoderV2;
+
+import "../ColonyDataTypes.sol";
+import "../IMetaColony.sol";
+import "./ColonyExtension.sol";
+
+
+contract ExtensionManager {
+  IMetaColony metaColony;
+
+  // [_extensionId][version] => resolver
+  mapping(bytes32 => mapping(uint256 => address)) resolvers;
+  // [_extensionId][version] => roles
+  mapping(bytes32 => mapping(uint256 => uint8[])) roles;
+  // [_extensionId][version][colony][domainId] => address
+  mapping(bytes32 => mapping(uint256 => mapping(address => mapping(uint256 => address payable)))) installations;
+
+  event ExtensionAdded(bytes32 indexed extensionId, uint256 version);
+  event ExtensionInstalled(bytes32 indexed extensionId, uint256 version, address indexed colony, uint256 indexed domainId);
+  event ExtensionDeprecated(bytes32 indexed extensionId, uint256 version, address indexed colony, uint256 indexed domainId);
+  event ExtensionUninstalled(bytes32 indexed extensionId, uint256 version, address indexed colony, uint256 indexed domainId);
+
+  constructor(address _metaColony) public {
+    metaColony = IMetaColony(_metaColony);
+  }
+
+  modifier onlyMetaColony() {
+    require(msg.sender == address(metaColony), "extension-manager-not-metacolony");
+    _;
+  }
+
+  modifier isAuthorized(address _colony, uint256 _permissionDomainId, uint256 _childSkillIndex, uint256 _domainId) {
+    require(IColony(_colony).userCanSetRoles(msg.sender, _permissionDomainId, _childSkillIndex, _domainId), "extension-manager-unauthorized");
+    _;
+  }
+
+  function addExtension(bytes32 _extensionId, uint256 _version, address _resolver, uint8[] memory _roles)
+    public
+    onlyMetaColony
+  {
+    require(_version == 0 || resolvers[_extensionId][_version - 1] != address(0x0), "extension-manager-bad-version");
+    require(resolvers[_extensionId][_version] == address(0x0), "extension-manager-already-added");
+    require(_resolver != address(0x0), "extension-manager-bad-resolver");
+
+    resolvers[_extensionId][_version] = _resolver;
+    roles[_extensionId][_version] = _roles;
+
+    emit ExtensionAdded(_extensionId, _version);
+  }
+
+  function installExtension(
+    bytes32 _extensionId,
+    uint256 _version,
+    address _colony,
+    uint256 _rootChildSkillIndex,
+    uint256 _permissionDomainId,
+    uint256 _childSkillIndex,
+    uint256 _domainId
+  )
+    public
+    isAuthorized(_colony, _permissionDomainId, _childSkillIndex, _domainId)
+  {
+    require(
+      installations[_extensionId][_version][_colony][_domainId] == address(0x0),
+      "extension-manager-already-installed"
+    );
+
+    ColonyExtension extension = new ColonyExtension();
```

Comment:
An extension deployed like this will have the default `install`, `uninstall`, `uninstallable` functions, regardless of what's defined in the resolver for that extension. Incoming transactions with those signatures will never reach the fallback function. This is exactly what I was talking about last Thursday. This has to be an EtherRouter.

Reply:
That is an excellent point.